pkg/ottl/ottlfuncs/func_delete_key.go

```diff
@@ -29,13 +32,39 @@ func createDeleteKeyFunction[K any](_ ottl.FunctionContext, oArgs ottl.Arguments
 	return deleteKey(args.Target, args.Key), nil
 }
 
-func deleteKey[K any](target ottl.PMapGetter[K], key string) ottl.ExprFunc[K] {
+func deleteKey[K any](target ottl.GetSetter[K], key string) ottl.ExprFunc[K] {
 	return func(ctx context.Context, tCtx K) (any, error) {
-		val, err := target.Get(ctx, tCtx)
+		val, err := anyToMap(ctx, tCtx, target)
 		if err != nil {
 			return nil, err
 		}
 		val.Remove(key)
-		return nil, nil
+		return nil, target.Set(ctx, tCtx, val)
+	}
+}
+
+var emptyMap = pcommon.Map{}
+
+func anyToMap[K any](ctx context.Context, tCtx K, target ottl.GetSetter[K]) (pcommon.Map, error) {
```

Comment:
Any advice where this should go into?

Reply:
It depends on https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/39416#discussion_r2056329367, if we create the new `PMapGetSetter` reusing the existing typed getters, this won't be needed anymore, otherwise it will probably be moved into the specific getter impl.